Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/stitch_borders.h

```diff
@@ -517,15 +520,62 @@ void stitch_borders(PolygonMesh& pmesh, const CGAL_PMP_NP_CLASS& np)
   std::vector< std::pair<halfedge_descriptor, halfedge_descriptor> > hedge_pairs_to_stitch;
 
   typedef typename GetVertexPointMap<PolygonMesh, CGAL_PMP_NP_CLASS>::const_type VPMap;
+  typedef typename VPMap::value_type Point_3;
   VPMap vpm = choose_param(get_param(np, internal_np::vertex_point),
                            get_const_property_map(vertex_point, pmesh));
+  bool stitch_along_ccs = choose_param(get_param(np, internal_np::do_stitch_per_cc),
+                           false);
 
-  internal::collect_duplicated_stitchable_boundary_edges(pmesh,
-    std::back_inserter(hedge_pairs_to_stitch),
-    internal::Less_for_halfedge<PolygonMesh, VPMap>(pmesh, vpm), vpm);
-
-  stitch_borders(pmesh, hedge_pairs_to_stitch);
-  internal::stitch_boundary_cycle_2(pmesh);
+  if(!stitch_along_ccs)
+  {
+    internal::collect_duplicated_stitchable_boundary_edges(pmesh,
+                                                           std::back_inserter(hedge_pairs_to_stitch),
+                                                           internal::Less_for_halfedge<PolygonMesh, VPMap>(pmesh, vpm), vpm);
+    
+    stitch_borders(pmesh, hedge_pairs_to_stitch);
+    internal::stitch_boundary_cycle_2(pmesh);
+  }
+  else
+  {
+    boost::unordered_map<typename boost::graph_traits<PolygonMesh>::face_descriptor,int> cc(num_faces(pmesh));
+    std::size_t num_component = CGAL::Polygon_mesh_processing::connected_components(pmesh, boost::make_assoc_property_map(cc));
+    
+    //add each border edge in map with cc
+    std::vector<std::vector<halfedge_descriptor> > border_edges_per_cc;
+    border_edges_per_cc.resize(num_component);
+    
+    BOOST_FOREACH(halfedge_descriptor hd, halfedges(pmesh))
+    {
+      halfedge_descriptor ohd = opposite(hd, pmesh);
+      if(is_border(ohd, pmesh))
+      {
+        border_edges_per_cc[cc[face(hd, pmesh)]].push_back(ohd);
+      }
+    }
+    // foreach cc make pair of corresponding edges.
+    typename boost::property_map< PolygonMesh,CGAL::vertex_point_t>::type vpmap
+        = get(CGAL::vertex_point, pmesh);
+    for(std::size_t i = 0; i < num_component; ++i)
+    {
+      for(int j = 0; j < static_cast<int>(border_edges_per_cc[i].size())-1; ++j)
+      {
+        halfedge_descriptor h1 = border_edges_per_cc[i].back();
+        border_edges_per_cc[i].pop_back();
+        Point_3 src1(get(vpmap, source(h1, pmesh))), tgt1(get(vpmap, target(h1, pmesh)));
+        BOOST_FOREACH(halfedge_descriptor h2, border_edges_per_cc[i])
+        {
+          Point_3 src2(get(vpmap, source(h2, pmesh))), tgt2(get(vpmap, target(h2, pmesh)));
+          if(src1 == tgt2 && src2 == tgt1)
+          {
+            hedge_pairs_to_stitch.push_back(std::make_pair(h1,h2));
+          }
+        }
+      }
+    }
```

Comment:
there is no handling of non-manifold edges here.
Maybe modifying the function `collect_duplicated_stitchable_boundary_edges` so that it takes as input a range of halfedges will make the job without to redo the collection job.